internal/operator/controller.go

```diff
@@ -0,0 +1,473 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	stanv1alpha1 "github.com/nats-io/nats-streaming-operator/pkg/apis/streaming/v1alpha1"
+	stancrdclient "github.com/nats-io/nats-streaming-operator/pkg/client/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	k8scorev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sfields "k8s.io/apimachinery/pkg/fields"
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	k8sschema "k8s.io/apimachinery/pkg/runtime/schema"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	k8sclient "k8s.io/client-go/kubernetes"
+	k8srestapi "k8s.io/client-go/rest"
+	k8scache "k8s.io/client-go/tools/cache"
+	k8sclientcmd "k8s.io/client-go/tools/clientcmd"
+)
+
+// Options for the operator.
+type Options struct {
+	// Namespace where the NATS Streaming Operator will
+	// be managing the clusters.
+	Namespace string
+
+	// NoSignals marks whether to enable the signal handler.
+	NoSignals bool
+}
+
+// Controller manages NATS Clusters running in Kubernetes.
+type Controller struct {
+	mu sync.Mutex
+
+	// Client to interact with Kubernetes resources.
+	kc k8sclient.Interface
+
+	// Client to interact with NATS Streaming Operator Kubernetes resources.
+	ncr stancrdclient.Interface
+
+	// clusters that the Operator is controlling.
+	clusters map[k8stypes.UID]*stanv1alpha1.NatsStreamingCluster
+
+	// opts is the set of options.
+	opts *Options
+
+	// quit stops the controller.
+	quit func()
+}
+
+func NewController(opts *Options) *Controller {
+	if opts == nil {
+		opts = &Options{}
+	}
+	return &Controller{
+		opts:     opts,
+		clusters: make(map[k8stypes.UID]*stanv1alpha1.NatsStreamingCluster),
+	}
+}
+
+// SetupClients takes the configuration and prepares the rest
+// clients that will be used to interact with the cluster objects.
+func (c *Controller) SetupClients(cfg *k8srestapi.Config) error {
+	kc, err := k8sclient.NewForConfig(cfg)
+	if err != nil {
+		return err
+	}
+	c.kc = kc
+
+	ncr, err := stancrdclient.NewForConfig(cfg)
+	if err != nil {
+		return err
+	}
+	c.ncr = ncr
+	return nil
+}
+
+// SetupSignalHandler enables handling process signals.
+func (c *Controller) SetupSignalHandler(ctx context.Context) {
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
+
+	for sig := range sigCh {
+		log.Debugf("Trapped '%v' signal", sig)
+
+		// If main context already done, then just skip
+		select {
+		case <-ctx.Done():
+			continue
+		default:
+		}
+
+		switch sig {
```

Comment:
Would you want a default operator to issue a trace statement that it'll be ignored?  I know you have the debug above, but it might someday help with debugging.

Reply:
oh good point, yes that sounds helpful for debugging